From f21a19fe4075f56d9a2412fa35cfca3f35d39ba7 Mon Sep 17 00:00:00 2001
From: Pavankumar Nandeshwar <pnandesh@qti.qualcomm.com>
Date: Tue, 15 Apr 2025 15:24:03 +0530
Subject: [PATCH] UPSTREAM: DP: wifi: ath12k: Handle irq registration for hybrid bus type

Handle irq registration for hybrid bus type, by adapting
the existing API to the new UD framework.

Patch-dependency: 0121-0004-wifi-ath12k-Configure-DP-interrupts-for-AHB.patch

Signed-off-by: Pavankumar Nandeshwar <pnandesh@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/ahb.c  |   6 +-
 drivers/net/wireless/ath/ath12k/pcic.c | 148 +++++++++++++++++--------
 drivers/net/wireless/ath/ath12k/pcic.h |   5 +
 3 files changed, 111 insertions(+), 48 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/ahb.c b/drivers/net/wireless/ath/ath12k/ahb.c
index f5033b4..6cc435c 100644
--- a/drivers/net/wireless/ath/ath12k/ahb.c
+++ b/drivers/net/wireless/ath/ath12k/ahb.c
@@ -645,8 +645,8 @@ ath12k_ahb_config_ext_irq(struct ath12k_base *ab,
 			/* For TX ring, ensure that the ring mask and the
 			 * tcl_to_wbm_rbm_map point to the same ring number.
 			 */
-			if (ring_mask->tx[i] &
-			    BIT(ab->hal.tcl_to_wbm_rbm_map[j].wbm_ring_num)) {
+			if ((j < DP_TCL_NUM_RING_MAX) && (ring_mask->tx[i] &
+			    BIT(ab->hal.tcl_to_wbm_rbm_map[j].wbm_ring_num))) {
 				irq_grp->irqs[num_irq++] =
 					wbm2host_tx_completions_ring1 - j;
 			}
@@ -804,7 +804,7 @@ static const struct ath12k_hif_ops ath12k_ahb_hif_ops_qcn6432 = {
         .map_service_to_pipe = ath12k_pcic_map_service_to_pipe,
 	.ce_irq_enable = ath12k_pcic_ce_irqs_enable,
 	.ce_irq_disable = ath12k_pcic_ce_irq_disable_sync,
-	.ext_irq_setup = ath12k_ahb_config_ext_irq,
+	.ext_irq_setup = ath12k_pcic_cfg_hybrid_ext_irq,
 };
 
 static irqreturn_t ath12k_userpd_irq_handler(int irq, void *data)
diff --git a/drivers/net/wireless/ath/ath12k/pcic.c b/drivers/net/wireless/ath/ath12k/pcic.c
index 3c77214..5b7403e 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -369,13 +369,18 @@ static void ath12k_pcic_ce_tasklet(struct tasklet_struct *t)
         ath12k_pcic_ce_irq_enable(ce_pipe->ab, ce_pipe->pipe_num);
 }
 
-static int ath12k_pcic_ext_config_gic_msi_irq(struct ath12k_base *ab,
-					      struct platform_device *pdev,
-					      struct msi_desc *msi_desc, int i)
+static
+int ath12k_pcic_ext_cfg_gic_msi_irq(struct ath12k_base *ab,
+				    int (*irq_handler)(struct ath12k_dp *dp,
+						       struct ath12k_ext_irq_grp *irq_grp,
+						       int budget),
+				    struct ath12k_dp *dp,
+				    struct msi_desc *msi_desc, int i)
 {
 	u32 user_base_data = 0, base_vector = 0, base_idx;
 	struct ath12k_ext_irq_grp *irq_grp;
 	struct ath12k_ahb *ab_ahb = ath12k_ab_to_ahb(ab);
+	struct platform_device *pdev = ab->pdev;
 	int j, budget, ret = 0, num_vectors = 0;
 	struct net_device *napi_ndev;
 	u8 userpd_id;
@@ -390,7 +395,9 @@ static int ath12k_pcic_ext_config_gic_msi_irq(struct ath12k_base *ab,
 
 	irq_grp = &ab->ext_irq_grp[i];
 	irq_grp->ab = ab;
+	irq_grp->dp = dp;
 	irq_grp->grp_id = i;
+	irq_grp->irq_handler = irq_handler;
 
 #if LINUX_VERSION_IS_GEQ(6,10,0)
 		irq_grp->napi_ndev = alloc_netdev_dummy(0);
@@ -439,9 +446,10 @@ static int ath12k_pcic_ext_config_gic_msi_irq(struct ath12k_base *ab,
 				       ath12k_pcic_ext_interrupt_handler, IRQF_SHARED,
 				       dp_irq_name[userpd_id][i], irq_grp);
 		if (ret) {
-			ath12k_err(ab, "failed request irq %d: %d\n", irq_idx, ret);
+			ath12k_warn(ab, "failed to request irq %d: %d\n", irq_idx, ret);
 			return ret;
 		}
+
 		ab->irq_num[irq_idx] = msi_desc->irq;
 		ab->ipci.dp_irq_num[vector] = msi_desc->irq;
 		ab->ipci.dp_msi_data[i] = msi_desc->msg.data;
@@ -852,6 +860,82 @@ static void ath12k_msi_msg_handler(struct msi_desc *desc, struct msi_msg *msg)
 	desc->msg.data = msg->data;
 }
 
+static
+int ath12k_pcic_msi_desc_assign_irq(struct ath12k_base *ab,
+				    int (*irq_handler)(struct ath12k_dp *dp,
+						       struct ath12k_ext_irq_grp *irq_grp,
+						       int budget),
+				    int base_vector, int num_vectors,
+				    struct platform_device *pdev, struct ath12k_dp *dp,
+				    int *k)
+{
+	struct msi_desc *msi_desc;
+	int ret = 0;
+	int i = 0;
+
+	msi_for_each_desc(msi_desc, &pdev->dev, MSI_DESC_ASSOCIATED) {
+		if (i < base_vector) {
+			i++;
+			continue;
+		}
+
+		if (i >= (base_vector + num_vectors) || *k >= ATH12K_EXT_IRQ_GRP_NUM_MAX)
+			break;
+
+		ret = ath12k_pcic_ext_cfg_gic_msi_irq(ab, irq_handler, dp,
+						      msi_desc, *k);
+		if (ret) {
+			ath12k_warn(ab, "failed to config ext msi irq %d\n", ret);
+			break;
+		}
+
+		(*k)++;
+		i++;
+	}
+
+	return ret;
+}
+
+int ath12k_pcic_cfg_hybrid_ext_irq(struct ath12k_base *ab,
+				   int (*irq_handler)(struct ath12k_dp *dp,
+					   struct ath12k_ext_irq_grp *irq_grp,
+					   int budget),
+				   struct ath12k_dp *dp)
+{
+	int ret, k = 0;
+	struct platform_device *pdev = ab->pdev;
+	struct ath12k_ahb *ab_ahb = ath12k_ab_to_ahb(ab);
+	int user_base_data, base_vector, num_vectors = 0;
+
+	if (ab_ahb->userpd_id != ATH12K_QCN6432_USERPD_ID_1 &&
+	    ab_ahb->userpd_id != ATH12K_QCN6432_USERPD_ID_2) {
+		ath12k_warn(ab, "ath12k userpd invalid %d\n", ab_ahb->userpd_id);
+		return -ENODEV;
+	}
+
+	msi_lock_descs(&pdev->dev);
+
+	ret = ath12k_pcic_get_user_msi_assignment(ab, "DP", &num_vectors,
+						  &user_base_data, &base_vector);
+	if (ret < 0) {
+		ath12k_warn(ab, "failed to fetch msi data for DP");
+		msi_unlock_descs(&pdev->dev);
+		return -EINVAL;
+	}
+
+	ret = ath12k_pcic_msi_desc_assign_irq(ab, irq_handler, base_vector, num_vectors,
+					      pdev, dp, &k);
+
+	/* This is needed to handle the case where the irqs need to be shared
+	 * when there are no enough msi lines available */
+	ret = ath12k_pcic_msi_desc_assign_irq(ab, irq_handler, base_vector, num_vectors,
+					      pdev, dp, &k);
+
+	msi_unlock_descs(&pdev->dev);
+
+	return ret;
+}
+
 int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab)
 {
 	int ret;
@@ -860,7 +944,7 @@ int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab)
 	struct ath12k_ahb *ab_ahb = ath12k_ab_to_ahb(ab);
 	bool ce_done = false;
 	int user_base_data, base_vector, num_vectors = 0;
-	int i = 0, j = 0, k = 0;
+	int i = 0, j = 0;
 
 	if (ab_ahb->userpd_id != ATH12K_QCN6432_USERPD_ID_1 &&
 	    ab_ahb->userpd_id != ATH12K_QCN6432_USERPD_ID_2) {
@@ -879,17 +963,18 @@ int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab)
 	}
 
 	msi_lock_descs(&pdev->dev);
-	//TODO: Need to optimize the below code to have one loop
-	msi_for_each_desc(msi_desc, &pdev->dev, MSI_DESC_ALL) {
-		ret = ath12k_pcic_get_user_msi_assignment(ab, "CE", &num_vectors,
-                                                         &user_base_data, &base_vector);
-                if (ret < 0)
-                        return ret;
 
+	ret = ath12k_pcic_get_user_msi_assignment(ab, "CE", &num_vectors,
+						  &user_base_data, &base_vector);
+	if (ret < 0)
+		return ret;
+	//TODO: Need to optimize the below code to have one loop
+	msi_for_each_desc(msi_desc, &pdev->dev, MSI_DESC_ASSOCIATED) {
 		if (i < base_vector) {
 			i++;
 			continue;
 		}
+
 		if (j < ab->hw_params->ce_count && i < (num_vectors + base_vector)) {
 			while(j < ab->hw_params->ce_count &&
 			      ath12k_ce_get_attr_flags(ab, j) & CE_ATTR_DIS_INTR) {
@@ -910,24 +995,9 @@ int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab)
 			}
 
 			j++;
-			if (j != ab->hw_params->ce_count)
-				ce_done = false;
-
-		} else {
-			ret = ath12k_pcic_get_user_msi_assignment(ab, "DP", &num_vectors,
-                                                                   &user_base_data, &base_vector);
-                        if (ret < 0)
-                                return ret;
-
-                        if (k >= num_vectors)
-                                break;
+			if (j >= ab->hw_params->ce_count)
+				ce_done = true;
 
-			ret = ath12k_pcic_ext_config_gic_msi_irq(ab, pdev, msi_desc, k);
-			if (ret) {
-				ath12k_warn(ab, "failed to config ext msi irq %d\n", ret);
-				return ret;
-			}
-			k++;
 		}
 		i++;
 	}
@@ -936,26 +1006,23 @@ int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab)
 	i = 0;
 
 	msi_lock_descs(&pdev->dev);
-	msi_for_each_desc(msi_desc, &pdev->dev, MSI_DESC_ALL) {
-		ret = ath12k_pcic_get_user_msi_assignment(ab, "CE", &num_vectors,
-							  &user_base_data, &base_vector);
-		if (ret < 0)
-                        return ret;
+	msi_for_each_desc(msi_desc, &pdev->dev, MSI_DESC_ASSOCIATED) {
+		if (ce_done)
+			break;
 
                 if (i < base_vector) {
                         i++;
                         continue;
 		}
 		if (i < (num_vectors + base_vector)) {
-			if (!ce_done  && j < ab->hw_params->ce_count) {
+			if (j < ab->hw_params->ce_count) {
 				while(j < ab->hw_params->ce_count &&
 				      ath12k_ce_get_attr_flags(ab, j) & CE_ATTR_DIS_INTR) {
 					j++;
 				}
 				if (j == ab->hw_params->ce_count) {
 					ce_done = true;
-					i++;
-					continue;
+					break;
 				}
 
 				ret = ath12k_pcic_config_gic_msi_irq(ab, pdev, msi_desc, j);
@@ -965,15 +1032,6 @@ int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab)
 				}
 				j++;
 			}
-		} else {
-			if (k >= ATH12K_EXT_IRQ_GRP_NUM_MAX)
-				break;
-			ret = ath12k_pcic_ext_config_gic_msi_irq(ab, pdev, msi_desc, k);
-			if (ret) {
-				ath12k_warn(ab, "failed to config ext msi irq %d\n", ret);
-				return ret;
-			}
-			k++;
 		}
 		i++;
 	}
diff --git a/drivers/net/wireless/ath/ath12k/pcic.h b/drivers/net/wireless/ath/ath12k/pcic.h
index 467fe1f..3048428 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.h
+++ b/drivers/net/wireless/ath/ath12k/pcic.h
@@ -77,5 +77,10 @@ int ath12k_pcic_ext_irq_config(struct ath12k_base *ab,
 						  struct ath12k_ext_irq_grp *irq_grp,
 						  int budget),
 			       struct ath12k_dp *dp);
+int ath12k_pcic_cfg_hybrid_ext_irq(struct ath12k_base *ab,
+				   int (*irq_handler)(struct ath12k_dp *dp,
+					   struct ath12k_ext_irq_grp *irq_grp,
+					   int budget),
+				   struct ath12k_dp *dp);
 void ath12k_pcic_free_ext_irq(struct ath12k_base *ab);
 #endif
-- 
2.34.1

